source/dbref.rst

```diff
@@ -0,0 +1,349 @@
+
+DBRefs have no relation to the deprecated DBPointer BSON type (i.e. type 0x0C).
+
+
+Examples of Valid DBRefs
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following examples are all valid DBRefs:
+
+.. code:: typescript
+
+  // Basic DBRef with only $ref and $id fields
+  { "$ref": "coll0", "$id": { "$oid": "60a6fe9a54f4180c86309efa" } }
+
+  // DBRef $id is not necessarily an ObjectId
+  { "$ref": "coll0", "$id": 1 }
+
+  // DBRef with optional $db field
+  { "$ref": "coll0", "$id": 1, "$db": "db0" }
+
+  // DBRef with extra, optional fields (with or without $db)
+  { "$ref": "coll0", "$id": 1, "$db": "db0", "foo": "bar" }
+  { "$ref": "coll0", "$id": 1, "foo": true }
+
+  // Extra field names have no inherent restrictions
+  { "$ref": "coll0", "$id": 1, "$foo": "bar" }
+  { "$ref": "coll0", "$id": 1, "foo.bar": 0 }
+
+Examples of Invalid DBRefs
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following examples are all invalid DBRefs:
+
+.. code:: typescript
+
+  // Required fields are omitted
+  { "$ref": "coll0" }
+  { "$id": { "$oid": "60a6fe9a54f4180c86309efa" } }
+
+  // Invalid types for $ref or $db
+  { "$ref": true, "$id": 1 }
+  { "$ref": "coll0", "$id": 1, "$db": 1 }
+
+  // Fields are out of order
+  { "$id": 1, "$ref": "coll0" }
+
+
+Implementing a DBRef Model
+--------------------------
+
+Drivers MAY provide a model class for encoding and/or decoding DBRef documents.
+For those drivers that do, this section defines expected behavior of that class.
+This section does not prohibit drivers from implementing additional
+functionality, provided it does not conflict with any of these guidelines.
+
+
+Constructing a DBRef model
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Drivers MAY provide an API for constructing a DBRef model directly from its
+constituent parts. If so:
+
+- Drivers MUST solicit a string value for ``$ref``.
+
+- Drivers MUST solicit an arbitary value for ``$id``. Drivers SHOULD NOT enforce
+  any restrictions on this value; however, this may be necessary if the driver
+  is unable to differentiate between certain BSON types (e.g. ``null``,
+  ``undefined``) and the parameter being unspecified.
+
+- Drivers SHOULD solicit an optional string value for ``$db``.
+
+- Drivers MUST require ``$ref`` and ``$db`` (if specfied) to be strings but
+  MUST NOT enforce any `naming restrictions`_ on the string values.
+
+- Drivers MAY solicit extra, optional fields.
+
+.. _naming restrictions: https://docs.mongodb.com/manual/reference/limits/#naming-restrictions
+
+Decoding a BSON document to a DBRef model
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Drivers MAY support explicit and/or implicit decoding. An example of explicit
+decoding might be a DBRef model constructor that takes a BSON document. An
+example of implicit decoding might be configuring the driver's BSON codec to
+automatically convert embedded documents that comply with the `DBRef Structure`_
+into a DBRef model.
+
+Drivers that provide implicit decoding MUST provide some way for applications to
```

Comment:
why is this a requirement?

Reply:
oh, I see this mentioned in the rationale section, so users can avoid changing the order of fields?

Reply:
It certainly helps address that, but I think it's also generally useful because DBRef really is just a convention and not a proper type that _needs_ a special class. I don't object to changing this to SHOULD if it's problematic as-is.

I'd be curious to know if this would require work for existing implementations, or if it's already supported. I suppose you can speak to Node, but that leaves C#, Java, Python, and Ruby.

cc: @ShaneHarvey, @p-mongo, @jyemin, @rstam

Reply:
Let me think about this some more. There currently seems to be the following issue:

- The dbref class in Ruby isn't derived from Hash.
- When Hash.from_bson is called, it may return a dbref based on the contents of bson document.
- This makes it possible for Hash.from_bson to not simply return different types but to return an unexpected type out of hierarchy.

Given that this is existing behavior I think some thought is need on my part as to how this might best be addressed.

Reply:
Are we anticipating a use case where, once dollars and dots are allowed in field names, applications may insert documents containing $ref and $id fields that they intend to simply be hashes, and not dbrefs? If yes it seems reasonable to require drivers to provide dbref-free decoding to support this use case.

Ruby would need to do some work to implement this since currently bson decoding isn't configurable, but there is precedent for similar functionality with extended json encoding (which can be canonical or relaxed) (https://www.rubydoc.info/github/mongodb/bson-ruby/BSON/Object#as_extended_json-instance_method) and I think it would be reasonable to implement something similar for bson decoding.

Reply:
> Are we anticipating a use case where, once dollars and dots are allowed in field names, applications may insert documents containing $ref and $id fields that they intend to simply be hashes, and not dbrefs? If yes it seems reasonable to require drivers to provide dbref-free decoding to support this use case.

It's quite possible for applications to start doing that with a 5.0-compat driver, but I wouldn't really anticipate it as a common behavior.

Since the spec will also require DBRef models allow access to any extra fields (e.g. `foo` for `{ $ref: "coll", $id: 1, "foo": 2 }`), I'm not worried about that either.

My reasoning is just that a DBRef model isn't a necessity, so giving (power?) users the ability to opt out of it is a good idea. Can we settle on changing this to SHOULD so Ruby (and others if needed) can punt on this for the time being?

Reply:
SHOULD works for me also.

Reply:
@p-mongo: Changed this in the latest commit.

Reply:
thanks for the explanation! given what you said and the conversation above the SHOULD makes sense to me. I will let @p-mongo resolve in case there's any further discussion to have here